ibis/common/scope.py

```diff
@@ -0,0 +1,107 @@
+    back. So in this case, we should not trust the data stored either, and go
+    on to execute this node. For simplicity, we update cache in this case as
+    well.
+
+"""
+from typing import Optional
+
+from ibis.expr.timecontext import TimeContextRelation, compare_timecontext
+from ibis.expr.typing import TimeContext
+
+
+def make_scope_item(op, result, timecontext: Optional[TimeContext]):
+    """make a scope item to be set in scope
+
+    Parameters
+    ----------
+    op: ibis.expr.operations.Node, key in scope.
+    result : scalar, pd.Series, pd.DataFrame, concrete data.
+    timecontext: Optional[TimeContext], time context associate with the
+    result.
+
+    Returns
+    -------
+    Dict, a key value pair that could merge into scope later
+    """
+    return {op: {'value': result, 'timecontext': timecontext}}
+
+
+def get_scope_item(scope, op, timecontext: Optional[TimeContext] = None):
+    """ Given a op and timecontext, get result from scope
+    Parameters
+    ----------
+    scope: collections.Mapping
+    a dictionary mapping :class:`~ibis.expr.operations.Node`
+    subclass instances to concrete data, and the time context associate
+    with it(if any).
+    op: ibis.expr.operations.Node, key in scope.
+    timecontext: Optional[TimeContext]
+
+    Returns
+    -------
+    result: scalar, pd.Series, pd.DataFrame
+    """
+    if op not in scope:
+        return None
+    # for ops without timecontext
+    if timecontext is None:
+        return scope[op].get('value', None)
+    else:
+        # For op with timecontext, ther are some ops cannot use cached result
+        # with a different (larger) timecontext to get the correct result.
+        # For example, a groupby followed by count, if we use a larger or
+        # smaller dataset from cache, we will probably get an error in result.
+        # Such ops with global aggregation, ops whose result is depending on
+        # other rows in result Dataframe, cannot use cached result with
+        # different time context to optimize calculation. These are time
+        # context sensitive operations. Since these cases are rare in
+        # acutal use case, we just enable optimization for all nodes for now.
+        old_timecontext = scope[op].get('timecontext', None)
+        if old_timecontext:
+            relation = compare_timecontext(timecontext, old_timecontext)
+            if relation == TimeContextRelation.SUBSET:
+                return scope[op].get('value', None)
+        else:
+            return scope[op].get('value', None)
+    return None
+
+
+def set_scope_item(scope, item):
+    """set scope_item in scope
+
+    Parameters
+    ----------
+    scope: collections.Mapping
+    item: Dict, generated by make_scope_item
+    """
+    for op, v in item.items():
+        # if get_scope returns a not None value, then data is already
+        # cached in scope and it is at least a greater range than
```

Comment:
Why the existing one is at least a greater range?

Reply:
If you execute a small window first the a large window, doesn't the scope contains the value for the small window first?

Reply:
Yes, in your case the scope contains the value for the small window. However, in `get_scope` we also passed in timecontext as a parameter. And with a larger timecontext as input, `get_scope` will return None according to its logic. Since the cached one is of a smaller time range and we shouldn't trust it.
Setting and getting shares similar logic of `test whether there is values cached with larget timecontext` and we use `get` method here in `set` to do such test